viewing_party/party.py

```diff
@@ -1,23 +1,147 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    movie = {}
 
+    if title and genre and rating:
+        movie["title"] = title
+        movie["genre"] = genre
+        movie["rating"] = rating
+        return movie
+    
+    else:
+        return None
+
+def add_to_watched(user_data, movie):
+    if movie not in user_data["watched"]:
+        user_data["watched"].append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    if movie not in user_data["watchlist"]:
+        user_data["watchlist"].append(movie)
+
+    return user_data
+    
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            add_to_watched(user_data, movie)
+            user_data["watchlist"].remove(movie)
+
+    return user_data
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0.0
+    
+    sum = 0.0
+
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    
+    avg_rating = sum / len(user_data["watched"])
+
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+    
+    counts = {}
+    most_watched_genre = user_data["watched"][0]["genre"]
+
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+
+        if genre not in counts:
+            counts[genre] = 0
+            
+        counts[genre] += 1
+
+        if counts[most_watched_genre] < counts[genre]:
+            most_watched_genre = genre
+
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    friends_movies = get_friends_movies_list_no_duplicates(user_data["friends"])
+    unique_movies = get_unique_list_1_movies(user_data["watched"], friends_movies)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    friends_movies = get_friends_movies_list_no_duplicates(user_data["friends"])        
+    unique_movies = get_unique_list_1_movies(friends_movies, user_data["watched"])
+
+    return unique_movies
+
+def get_friends_movies_list_no_duplicates(friends_list):
+    friends_movies = []
+
+    for friend in friends_list:
+        for movie in friend["watched"]:
+            if movie not in friends_movies:
+                friends_movies.append(movie)
+
+    return friends_movies

```

Comment:
I think that overall, getting a collection of all the unique movies for the friends is a great idea. And this function works well. 

One small concern though is that the `not in` step on line 92 does technically make this a triply nested loop which is not ideal. An obvious replacement would be to use sets as they are implemented using Hash maps and allow us to add while avoiding duplicate items in constant time. We do have a slight drawback in the fact that sets do not work with mutable elements (We couldn't add dictionaries to a set for example). 

With this in mind, we could create a set of movie titles that could be used later. 

*Note that this particular approach would require us to modify the helper function and potentially the functions that call this one as well!